File: tic-tac-toe/java/src/main/java/com/example/tictactoe/game/Game.java
package com.example.tictactoe.game;

import com.example.tictactoe.exception.InvalidMoveException;
import com.example.tictactoe.model.GameStatus;
import com.example.tictactoe.model.Mark;
import com.example.tictactoe.model.Player;

/**
 * The game engine: validates whose turn it is, delegates board writes, then advances the state.
 * It is intentionally single-threaded/turn-based; callers serialize moves like players do in real play.
 */
public class Game {

    private final Board board;
    private final Player xPlayer;
    private final Player oPlayer;
    private Player currentPlayer;
    private GameStatus status = GameStatus.IN_PROGRESS;

    public Game(Player xPlayer, Player oPlayer) {
        this(Board.DEFAULT_SIZE, xPlayer, oPlayer);
    }

    public Game(int size, Player xPlayer, Player oPlayer) {
        if (xPlayer.mark() != Mark.X || oPlayer.mark() != Mark.O) {
            throw new IllegalArgumentException("Game requires one X player and one O player");
        }
        this.board = new Board(size);
        this.xPlayer = xPlayer;
        this.oPlayer = oPlayer;
        this.currentPlayer = xPlayer;
    }

    public GameStatus makeMove(Player player, int row, int col) {
        if (status != GameStatus.IN_PROGRESS) {
            throw new InvalidMoveException("Game is already over with status " + status);
        }
        if (!player.equals(currentPlayer)) {
            throw new InvalidMoveException("It is " + currentPlayer.name() + "'s turn");
        }

        board.placeMark(row, col, player.mark());
        if (board.hasWinningLine(player.mark())) {
            status = player.mark() == Mark.X ? GameStatus.X_WON : GameStatus.O_WON;
        } else if (board.isFull()) {
            status = GameStatus.DRAW;
        } else {
            currentPlayer = currentPlayer.equals(xPlayer) ? oPlayer : xPlayer;
        }
        return status;
    }

    public Board getBoard() {
        return board;
    }

    public Player getCurrentPlayer() {
        return currentPlayer;
    }

    public GameStatus getStatus() {
        return status;
    }
}

File: tic-tac-toe/python/tictactoe/game.py
"""The Tic Tac Toe game engine."""

from __future__ import annotations

from .exceptions import InvalidMoveError
from .models import Board, GameStatus, Mark, Player


class Game:
    """Validate turns, delegate board writes, then transition game status.

    This class is intentionally single-threaded/turn-based. A caller supplies one move at a time,
    so locks would add noise without improving the correctness of the MVP.
    """

    def __init__(self, x_player: Player, o_player: Player, size: int = Board.DEFAULT_SIZE) -> None:
        if x_player.mark is not Mark.X or o_player.mark is not Mark.O:
            raise ValueError("Game requires one X player and one O player")
        self.board = Board(size)
        self._x_player = x_player
        self._o_player = o_player
        self.current_player = x_player
        self.status = GameStatus.IN_PROGRESS

    def make_move(self, player: Player, row: int, col: int) -> GameStatus:
        if self.status is not GameStatus.IN_PROGRESS:
            raise InvalidMoveError(f"Game is already over with status {self.status.value}")
        if player != self.current_player:
            raise InvalidMoveError(f"It is {self.current_player.name}'s turn")

        self.board.place_mark(row, col, player.mark)
        if self.board.has_winning_line(player.mark):
            self.status = GameStatus.X_WON if player.mark is Mark.X else GameStatus.O_WON
        elif self.board.is_full():
            self.status = GameStatus.DRAW
        else:
            self.current_player = self._o_player if self.current_player == self._x_player else self._x_player
        return self.status
